### ISSUE TITLE
ListenMessages Authorization Rule Resulting in MessagingGatewayConflict

### ISSUE THREAD

**Issue by wjohnson** [microsoft/Purview-ADB-Lineage-Solution-Accelerator#120]
**Describe the bug**
When doing the connector only deployment, it is giving an event hub `MessagingGatewayConflict` error for the ListenMessages Authorization Rules

**To Reproduce**
1. Run the connector only deployment from release 2.1 branch

**Expected behavior**
It should deploy the connector completely.

**Logs**

```json
{
    "status": "Failed",
    "error": {
        "code": "DeploymentFailed",
        "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
        "details": [
            {
                "code": "Conflict",
                "message": {
                    "status": "failed",
                    "error": {
                        "code": "ResourceDeploymentFailure",
                        "message": "The resource operation completed with terminal provisioning state 'failed'.",
                        "details": [
                            {
                                "code": "Failed",
                                "message": "ARM-MSDeploy Deploy Failed: 'System.Threading.ThreadAbortException: Thread was being aborted.
   at System.Collections.Generic.SortedList`2..ctor()
   at Microsoft.Web.Deployment.DeploymentBaseContext..ctor(DeploymentBaseOptions baseOptions, DeploymentObject sourceObject)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObjectPrivate(DeploymentProviderContext providerContext, DeploymentBaseOptions baseOptions, DeploymentObject sourceObject, String serverVersion)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentProviderOptions providerOptions, DeploymentBaseOptions baseOptions)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObject(String provider, String path, DeploymentBaseOptions baseOptions)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentWellKnownProvider provider, String path, DeploymentBaseOptions baseOptions)
   at Microsoft.Web.Deployment.WebApi.AppGalleryPackage.Deploy(String deploymentSite, String siteSlotId, Boolean doNotDelete)
   at Microsoft.Web.Deployment.WebApi.DeploymentController.&lt;DownloadAndDeployPackage&gt;d__25.MoveNext()'"
                            }
                        ]
                    }
                }
            },
            {
                "code": "Conflict",
                "message": {
                    "error": {
                        "code": "MessagingGatewayConflict",
                        "message": "<Error><Code>409</Code><Detail> TrackingId:abac1930-22e0-41b2-9af2-a06fbeaabae9_G23, SystemTracker:eventhubnsgh1142dhbd.servicebus.windows.net:$tenants/eventhubnsgh1142dhbd, Timestamp:2022-11-26T17:48:05</Detail></Error>"
                    }
                }
            }
        ]
    }
}
```

**Desktop (please complete the following information):**
 - OS: Cloudshell with Bash
 - OpenLineage Version: N/A
 - Databricks Runtime Version: N/A
 - Cluster Type: N/A
 - Cluster Mode: N/A
 - Using Credential Passthrough: N/A

**Additional context**
N/A

**Comment by wjohnson** [comment]
@TravisHilbert have you seen this error before?

**Comment by wjohnson** [comment]
I can't find it in the documented ARM errors: https://learn.microsoft.com/en-us/azure/event-hubs/resource-manager-exceptions

**Comment by SumitRauniyarNeu** [comment]
![image](https://user-images.githubusercontent.com/91178442/205058300-ea8d61a6-4a36-4244-a29a-ea8f79e980b8.png)
What input do I give here? Unable to deploy connector only because of this.

**Comment by wjohnson** [comment]
> ![image](https://user-images.githubusercontent.com/91178442/205058300-ea8d61a6-4a36-4244-a29a-ea8f79e980b8.png) What input do I give here? Unable to deploy connector only because of this.

Hi, @SumitRayniyarNeu - That parameter is being removed on the latest release (2.2.0) but if you need to stick with 2.1.x, you should provide the Purview Event Hub connection string.

**Comment by SumitRauniyarNeu** [comment]
Confirming the same @himanshugarg1ext  @TravisHilbert  @wjohnson 
This issue has not resolved even in the latest 2.2 release. Unable to deploy this package.

MessagingGatewayConflict",
      "message": "<Error><Code>409</Code>

**Comment by wjohnson** [comment]
Thank you for the feedback.

If you run the deployment twice, you will not receive the error the second
time.

Please run the deployment again with the same parameters and it will do an
incremental update.

Please let us know if running the deployment a second time does not resolve
the issue (at least temporarily).

**Comment by himanshugarg1ext** [comment]
@wjohnson @TravisHilbert yes I tried this in the day, it actually worked.

However, I am running into another problem with the redeployment approach.
We want to use our existing Key Vault (instead of deploying a new one with Access Policies) for following reasons:
1. The existing Key Vault is configured with RBAC due to organizational policies.
2. The existing Key Vault has already the ClientID and ClientSecret stored in it.
3. We can use the same Key Vault which is configured in our ADB workspace, and can use the Function Key as a _secret_, which will be injected in our KV with this template after creating the function (the same way it is ding to a new KV currently), This way we won't have to create another AKV backed secret scope in our ADB workspace for this sole purpose.

In order for that to work, I have added a RoleAssignment module within this template (to assign Key Vault Secret User Role to my KV since we have RBAC instead of Access Policy) which throws an error _on redeployment_ with this:

{
    "status": "Failed",
    "error": {
        "code": "RoleAssignmentUpdateNotPermitted",
        "message": "Tenant ID, application ID, principal ID, and scope are not allowed to be updated."
    }
}


Retry option takes away the original error but presents with another problem in our case. :/

**Comment by himanshugarg1ext** [comment]
I'm wondering how did it stop working on R/2.0. It worked perfectly fine earlier when I was trying this out in our PoC on this accelerator. Now when we want to productionize it, it breaks.
Has there been any recent changes on R/2.0 (can't see them though)?

**Comment by wjohnson** [comment]
> I'm wondering how did it stop working on R/2.0. It worked perfectly fine earlier when I was trying this out in our PoC on this accelerator. Now when we want to productionize it, it breaks. Has there been any recent changes on R/2.0 (can't see them though)?

There have been no changes to the ARM template that would have affected it.  We are working with some internal teams to help figure out why the issue occurs intermittently.

> We can use the same Key Vault which is configured in our ADB workspace, and can use the Function Key as a secret, which will be injected in our KV with this template after creating the function (the same way it is ding to a new KV currently), This way we won't have to create another AKV backed secret scope in our ADB workspace for this sole purpose.

I would encourage you to keep a separate key vault to be used just for this accelerator.  Databricks secrets are not actually that secret and can be extracted very easily if a user is on an interactive cluster or job cluster.  The client id and secret you're providing have the Data Curator role in Purview. As a result, in your situation, an app developer on Databricks can extract the secret from the Databricks secret and then begin manipulating assets in Microsoft Purview.

If you keep the app id and secret in its own separate key vault, you'll be able to avoid the above situation and keep your secrets actually secret.

**Comment by SumitRauniyarNeu** [comment]
@wjohnson Can you suggest how can I push lineage when I perform merge?

**Comment by wjohnson** [comment]
Hi, all!  This is now resolved with the latest releases for release/2.1 and release/2.2 branches.

See https://github.com/microsoft/Purview-ADB-Lineage-Solution-Accelerator/releases/tag/2.1.2 and https://github.com/microsoft/Purview-ADB-Lineage-Solution-Accelerator/releases/tag/2.2.2
